File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        int ans = 0, start = -1;
        Map<Character, Integer> map = new HashMap<>();
        
        for (int i = 0; i < s.length(); i++) {
            char c = s.charAt(i);
            
            if (map.containsKey(c)) {
                start = Math.max(start, map.get(c));
            }
            
            map.put(c, i);
            ans = Math.max(ans, i - start);
        }
        
        return ans;
    }
}

File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        seen = {}
        left = ma = 0
        for i, ch in enumerate(s):
            if ch in seen and seen[ch] >= left:
                left = seen[ch] + 1          
            ma = max(ma, i - left + 1)       
            seen[ch] = i                     
        return ma
                 